syntax = "proto3";

package convergedcomputing.org.grpc.v1;

import "google/protobuf/timestamp.proto";

option go_package = "github.com/converged-computing/rainbow/pkg/api/v1";

// RainbowSchedulerService provides API endpoints for interacting with the central scheduler service
service RainbowScheduler {

  // SaveMetric saves a metric to the database
  rpc SaveMetric(SaveMetricRequest) returns (SaveMetricResponse);

  // Register cluster - request to register a new cluster
  rpc Register(RegisterRequest) returns (RegisterResponse);

  // Register cluster - request to register a new cluster
  rpc RegisterSubsystem(RegisterRequest) returns (RegisterResponse);

  // Job Submission - request for submitting a job to a named cluster
  rpc SubmitJob(SubmitJobRequest) returns (SubmitJobResponse);

  // Update State - allow a cluster to provide state metadata
  // This is intended for use by a selection algorithm
  rpc UpdateState(UpdateStateRequest) returns (UpdateStateResponse);

  // Request Job - ask the rainbow scheduler for up to max jobs
  rpc ReceiveJobs(ReceiveJobsRequest) returns (ReceiveJobsResponse);

  // Accept Jobs - accept some number of jobs
  rpc AcceptJobs(AcceptJobsRequest) returns (AcceptJobsResponse);
}

// RegisterRequest registers a cluster to the scheduler service
// The shared secret is required to validate the request
message RegisterRequest {
  string name = 1;
  string secret = 2;

  // JGF json of nodes (needs to be read in with jsongraph-go)
  string nodes = 3;
  string subsystem = 4;
  google.protobuf.Timestamp sent = 5;
}

message SaveMetricRequest {
  string name = 1;
  string value = 2;
  map<string, string> metadata = 3;
}

message SaveMetricResponse {
  enum ResultType {
    SAVE_METRIC_SUCCESS = 0;
    SAVE_METRIC_ERROR = 1;
  }
  ResultType status = 1;
}

// UpdateStateRequests allows a cluster to set arbitrary metadata
// for its state. State metadata is used for selection algorithms
message UpdateStateRequest {
  string cluster = 1;
  string secret = 2;

  // We are generous that the payload can be a flat
  // set of key value pairs, and will be parsed into
  // types within the graph database
  string payload = 3;
}

message UpdateStateResponse {
  enum ResultType {
    UPDATE_STATE_UNSPECIFIED = 0;
    UPDATE_STATE_PARTIAL = 1;
    UPDATE_STATE_SUCCESS = 2;
    UPDATE_STATE_ERROR = 3;
  }
  ResultType status = 1;
}


// SubmitJobRequest takes a job name, cluster name
// and requires the cluster token. Since we want to be generic,
// we currently accept nodes, tasks, and the command
message SubmitJobRequest {
  string name = 1;
  repeated Cluster clusters = 2;
  string jobspec = 3;
  string select_algorithm = 4;
  map<string,string> select_options = 5;
  bool satisfy_only = 6;
  google.protobuf.Timestamp sent = 7;
  int64 jobid = 8;

  message Cluster {
    string name = 1;
    string token = 2;
  }
}

// RequestJobsRequest is used by a cluster (or other entity that can run jobs)
// to get back a maximum of N jobs. This will (and should) eventually
// support filters / sorting criteria, but now we just take the top
// N from a query. The response includes the ids that are accepted
message ReceiveJobsRequest {
  string cluster = 1;

  // cluster secret given on registration
  // No other cluster or user can take a cluster's jobs!
  string secret = 2;

  // Max jobs to request
  int32 maxJobs = 3;
  google.protobuf.Timestamp sent = 7;
}

// N from a query. The response includes the ids that are accepted
message AcceptJobsRequest {
  string cluster = 1;
  string secret = 2;
  repeated int32 jobids = 3;
  google.protobuf.Timestamp sent = 4;
}

// Register Response
message RegisterResponse {

  // Registration statuses
  enum ResultType {
    REGISTER_UNSPECIFIED = 0;
    REGISTER_SUCCESS = 1;
    REGISTER_ERROR = 2;
    REGISTER_DENIED = 3;
    REGISTER_EXISTS = 4;
  }
  string request_id = 1;

  // The "token" is given to clients (or this cluster) to submit jobs
  // In the future this should be able to be refreshed
  string token = 2;

  // The "secret" is for this cluster to receive them.
  string secret = 3;
  ResultType status = 4;
}

// Submit Job Response
message SubmitJobResponse {

  // Enum to represent the result types of the operation.
  enum ResultType {
    SUBMIT_UNSPECIFIED = 0;
    SUBMIT_SUCCESS = 1;
    SUBMIT_ERROR = 2;
    SUBMIT_DENIED = 3;
  }
  string request_id = 1;
  int32 jobid = 2;
  string cluster = 3;
  ResultType status = 4;
  // A SatisfyOnly request can return the entire selected set
  repeated string clusters = 5;
}

// RequestJobsResponse
message ReceiveJobsResponse {

  // Enum to represent the result types of the operation.
  enum ResultType {
    REQUEST_JOBS_NORESULTS = 0;
    REQUEST_JOBS_SUCCESS = 1;
    REQUEST_JOBS_ERROR = 2;
    REQUEST_JOBS_DENIED = 3;
  }
  string request_id = 1;
  map<int32, string> jobs = 2;
  //  repeated int32 accepted_jobids = 2;
  ResultType status = 3;
}

// Accept Jobs Response
message AcceptJobsResponse {

  enum ResultType {
    RESULT_TYPE_UNSPECIFIED = 0;
    RESULT_TYPE_PARTIAL = 1;
    RESULT_TYPE_SUCCESS = 2;
    RESULT_TYPE_ERROR = 3;
  }
  ResultType status = 1;
}
